Snowflake ignore case (#7698)
fishtown-analytics/docs.getdbt.com: website/docs/reference/project-configs/quoting.md

  database: true | false
  schema: true | false
  identifier: true | false
  snowflake_ignore_case: true | false  # Fusion-only config. Aligns with Snowflake's session parameter QUOTED_IDENTIFIERS_IGNORE_CASE behavior. 

>                                        # Ignored by dbt Core and other adapters.

Reviewer: Hiya @runleonarun 

This is for my personal knowledge - I'm assuming this `snowflake_ignore_case: true | false  # Fusion-only config. Aligns with Snowflake's session parameter QUOTED_IDENTIFIERS_IGNORE_CASE behavior. 
                                       # Ignored by dbt Core and other adapters` can't be a snippet as it's part of a code block.

Kind Regards
Natalie

Author: Yea snippets within code don't work :(

</File>

## Recommendations

### Snowflake

If you're using Snowflake, we recommend:

> - Setting all quoting configs to `False` in your dbt_project.yml to avoid quoting model and column names unnecessarily and to help prevent case sensitivity issues.

Reviewer: ```suggestion
- Setting all quoting configs to `False` in your [`dbt_project.yml`](/reference/dbt_project.yml) to avoid quoting model and column names unnecessarily and to help prevent case sensitivity issues.
```

  - Setting all quoting configs to `False` also means you cannot use reserved words as identifiers, such as model or table names. We recommend you avoid using these reserved words anyway.  

> - If you're using Fusion and your Snowflake environment sets the session parameter `QUOTED_IDENTIFIERS_IGNORE_CASE = true` (for example, in an orchestrator or pre-hook), you should also enable quoting and `snowflake_ignore_case` in your dbt_project.yml to preserve the exact case of database, schema, and identifier:

Reviewer: ```suggestion
- If you're using Fusion and your Snowflake environment sets the session parameter `QUOTED_IDENTIFIERS_IGNORE_CASE = true` (for example, in an orchestrator or pre-hook), you should also enable quoting and `snowflake_ignore_case` in your `dbt_project.yml` to preserve the exact case of database, schema, and identifier:
```

  ```yml
  quoting:
    database: true
    schema: true
    identifier: true
    snowflake_ignore_case: true  # Fusion-only config. Aligns with Snowflake's session parameter QUOTED_IDENTIFIERS_IGNORE_CASE behavior. 
                                 # Ignored by dbt Core and other adapters.
  ```

>   Setting `snowflake_ignore_case: true` ensures that dbt compiles column and identifier names to match Snowflake’s behavior at runtime, preserving parity between compile-time and runtime logic. Without this, you may encounter "column not found" errors.

Reviewer: ```suggestion
  Setting `snowflake_ignore_case: true` ensures that dbt compiles column and identifier names match Snowflake’s behavior at runtime, preserving parity between compile-time and runtime logic. Without this, you may encounter "column not found" errors.
```

:::info Quoting a source
If a Snowflake source table uses a quoted database, schema, or table identifier, you can configure this in the source.yml file. Refer to [configuring quoting](/reference/resource-properties/quoting) for more information.
:::

#### Explanation

> Unlike most databases (which lowercase unquoted identifiers), Snowflake uppercases them. But when identifiers are quoted, case is preserved. This creates problems when dbt compiles a model with quoted, lowercase identifiers, and the query later references those identifiers without quotes, triggering uppercasing and a mismatch.

Reviewer: ```suggestion
Unlike most databases (which lowercases unquoted identifiers), Snowflake uppercases them. But when identifiers are quoted, case is preserved. This creates problems when dbt compiles a model with quoted, lowercase identifiers, and the query later references those identifiers without quotes, triggering uppercasing and a mismatch.
```

> Unlike most databases (which lowercases unquoted identifiers), Snowflake uppercases them. But when identifiers are quoted, case is preserved. This creates problems when dbt compiles a model with quoted, lowercase identifiers, and the query later references those identifiers without quotes, triggering uppercasing and a mismatch.

Reviewer 2: @runleonarun Both the original _and_ there rewrite are difficult for me to understand, largely because there's multiple moving parts to dbt's behavior. For this specific rewrite, I had a hard time understanding what "the query" was referring to.

Here's an attempt at a minimal update to the docs that explains the key rationale:

> Whereas most databases will _lowercase_ unquoted identifiers, Snowflake will _uppercase_ unquoted identifiers. If a model name is lowercased _and quoted_, then it cannot be referred to without quotes! dbt skips quoting on Snowflake so lowercase model names work seamlessly in downstream queries and BI tools without worrying about case or quotes.

This is driven by dbt's convention of lowercase file names and lowercase model names. If the dbt convention was opposite in this regard, then the default quoting rules would also be inverted.

Here's a longer explanation that tries to explain the key nuances:

> dbt skips quoting on Snowflake so lowercase model names work seamlessly in downstream queries and BI tools without worrying about case or quotes.
> 
> Unlike most databases (which lowercase unquoted identifiers), Snowflake uppercases them. When identifiers are quoted, Snowflake preserves their case and makes them case-sensitive. This means a table created with quoted, lowercase identifiers must always be referenced with quotes and the exact same case -- something that can easily break downstream queries in BI tools or ad-hoc SQL. Because dbt conventions use lowercase model and file names, quoting them on Snowflake would cause this problem; if dbt instead used uppercase names by convention, the safe defaults for Snowflake and other databases would be reversed.

Author: Hey @dbeatty10 I used your more detailed explanation and tweaked it.

> Unlike most databases (which lowercases unquoted identifiers), Snowflake uppercases them. But when identifiers are quoted, case is preserved. This creates problems when dbt compiles a model with quoted, lowercase identifiers, and the query later references those identifiers without quotes, triggering uppercasing and a mismatch.

Author: ```suggestion
dbt skips quoting on Snowflake so lowercase model names work seamlessly in downstream queries and BI tools without worrying about case or quotes.

Unlike most databases (which lowercase unquoted identifiers), Snowflake uppercases them. When you quote identifiers, Snowflake will preserve their case and make them case-sensitive. This means when you create a table with quoted, lowercase identifiers, the table should always be referenced with quotes and use the exact same case, which can easily break downstream queries in BI tools or ad-hoc SQL. 

Because dbt conventions use lowercase model and file names, quoting them in Snowflake risks breaking downstream queries in BI tools or ad-hoc SQL. If dbt instead used uppercase names by convention, the safe defaults for other databases would be at risk of breaking downstream queries.
```

## Recommendation

### Snowflake

If you're using Snowflake, we recommend:

- Setting all quoting configs to `False` in your dbt_project.yml to avoid quoting model and column names unnecessarily and to help prevent case sensitivity issues.
  - Setting all quoting configs to `False` also means you cannot use reserved words as identifiers, such as model or table names. We recommend you avoid using these reserved words anyway.
- If your Snowflake environment sets the session parameter `QUOTED_IDENTIFIERS_IGNORE_CASE = true` (for example, in an orchestrator or pre-hook), you must also set `snowflake_ignore_case: true` and `identifier: true` in your dbt_project.yml.

  Setting `snowflake_ignore_case: true` ensures that dbt compiles column and identifier names to match Snowflake’s behavior at runtime, preserving parity between compile-time and runtime logic. Without this, you may encounter "column not found" errors.

> ::: Note

Author: ```suggestion
:::info Quoting a source
```

> If a Snowflake source table uses a quoted database, schema, or table identifier, you can configure this in the source.yml file. [Refer to configuring quoting for more info](/reference/resource-properties/quoting).

Author: ```suggestion
If a Snowflake source table uses a quoted database, schema, or table identifier, you can configure this in the source.yml file. Refer to [configuring quoting](/reference/resource-properties/quoting) for more information.
```
